File: src/services/mirofish/socialGraph.ts
/**
 * socialGraph.ts — OASIS social network graph
 *
 * Generates a realistic follow network using the Barabasi-Albert
 * preferential attachment model.
 * TypeScript reimplementation of OASIS social_graph.py.
 */

export class SocialGraph {
  private following = new Map<string, Set<string>>()  // agentId -> set of IDs being followed
  private followers = new Map<string, Set<string>>()  // agentId -> set of follower IDs

  constructor(agentIds: string[]) {
    for (const id of agentIds) {
      this.following.set(id, new Set())
      this.followers.set(id, new Set())
    }
  }

  /**
   * Generate a social graph using the Barabasi-Albert model.
   * m: number of existing nodes a new node connects to (default 2).
   * Nodes with more followers have higher probability of receiving new follows.
   */
  static generate(agentIds: string[], m = 2): SocialGraph {
    const graph = new SocialGraph(agentIds)
    if (agentIds.length < 2) return graph

    const initial = agentIds.slice(0, Math.min(m + 1, agentIds.length))
    for (const a of initial) {
      for (const b of initial) {
        if (a !== b) graph.follow(a, b)
      }
    }

    for (const newNode of agentIds.slice(initial.length)) {
      for (const target of graph._preferentialAttachment(newNode, m)) {
        graph.follow(newNode, target)
      }
    }

    return graph
  }

  private _preferentialAttachment(exclude: string, m: number): string[] {
    const pool: string[] = []
    for (const [id, fset] of this.followers) {
      if (id === exclude) continue
      const weight = fset.size + 1
      for (let i = 0; i < weight; i++) pool.push(id)
    }
    // Fisher-Yates shuffle (Math.random().sort() has distribution bias)
    for (let i = pool.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1))
      ;[pool[i], pool[j]] = [pool[j], pool[i]]
    }
    const targets = new Set<string>()
    for (const c of pool) {
      if (targets.size >= m) break
      targets.add(c)
    }
    return [...targets]
  }

  follow(from: string, to: string): void {
    this.following.get(from)?.add(to)
    this.followers.get(to)?.add(from)
  }

  unfollow(from: string, to: string): void {
    this.following.get(from)?.delete(to)
    this.followers.get(to)?.delete(from)
  }

  getFollowing(agentId: string): string[] {
    return [...(this.following.get(agentId) ?? [])]
  }

  getFollowers(agentId: string): string[] {
    return [...(this.followers.get(agentId) ?? [])]
  }

  followerCount(agentId: string): number {
    return this.followers.get(agentId)?.size ?? 0
  }

  allAgentIds(): string[] {
    return [...this.following.keys()]
  }
}

File: bot/modules/social_graph.py
"""
social_graph.py — OASIS Social Network Graph

Generates a realistic follow network using the Barabási-Albert preferential attachment model.
Reimplements OASIS social_graph.py in Python.
"""
from __future__ import annotations

import random


class SocialGraph:
    def __init__(self, agent_ids: list[str]) -> None:
        self._following: dict[str, set[str]] = {a: set() for a in agent_ids}
        self._followers: dict[str, set[str]] = {a: set() for a in agent_ids}

    @classmethod
    def generate(cls, agent_ids: list[str], m: int = 2) -> "SocialGraph":
        """
        Generate a social graph using the Barabási-Albert model.
        m: number of existing nodes each new node connects to (default 2).
        Nodes with more followers have a higher probability of receiving new follows.
        """
        graph = cls(agent_ids)
        if len(agent_ids) < 2:
            return graph

        # Initial clique (first m+1 nodes follow each other)
        initial = agent_ids[: min(m + 1, len(agent_ids))]
        for a in initial:
            for b in initial:
                if a != b:
                    graph.follow(a, b)

        # Add remaining nodes sequentially — preferential attachment
        for new_node in agent_ids[len(initial) :]:
            for target in graph._preferential_attachment(new_node, m):
                graph.follow(new_node, target)

        return graph

    def _preferential_attachment(self, exclude: str, m: int) -> list[str]:
        """Select m connection targets with weights proportional to follower count."""
        pool: list[str] = []
        for agent_id, follower_set in self._followers.items():
            if agent_id == exclude:
                continue
            pool.extend([agent_id] * (len(follower_set) + 1))
        random.shuffle(pool)
        targets: set[str] = set()
        for candidate in pool:
            if len(targets) >= m:
                break
            targets.add(candidate)
        return list(targets)

    # ── Follow Operations ──────────────────────────────────────────────────────

    def follow(self, from_id: str, to_id: str) -> None:
        self._following.setdefault(from_id, set()).add(to_id)
        self._followers.setdefault(to_id, set()).add(from_id)

    def unfollow(self, from_id: str, to_id: str) -> None:
        self._following.get(from_id, set()).discard(to_id)
        self._followers.get(to_id, set()).discard(from_id)

    # ── Queries ────────────────────────────────────────────────────────────────

    def get_following(self, agent_id: str) -> list[str]:
        return list(self._following.get(agent_id, set()))

    def get_followers(self, agent_id: str) -> list[str]:
        return list(self._followers.get(agent_id, set()))

    def follower_count(self, agent_id: str) -> int:
        return len(self._followers.get(agent_id, set()))

    def all_agent_ids(self) -> list[str]:
        return list(self._following.keys())
